src/main/kotlin/io/icure/md/client/apis/impl/HealthcareElementApiImpl.kt

```diff
@@ -95,15 +106,80 @@ class HealthcareElementApiImpl(private val api: MedTechApi) : HealthcareElementA
+        val currentUser = medTechApi.baseUserApi.getCurrentUser()
+        val dataOwnerId = currentUser.dataOwnerId()
+
+        if (!healthcareElement.systemMetaData!!.delegations.keys.any { it == dataOwnerId }) {
+            throw IllegalStateException("DataOwner $dataOwnerId does not have the right to access it ${healthcareElement.id}")
+        }
+
+        if (healthcareElement.systemMetaData.delegations.keys.any { it == delegateTo }) {
+            return healthcareElement
+        }
+
+        val ccHealthElement = healthElementCryptoConfig(localCrypto)
+
+        return healthcareElement.toHealthcareElementDto().let { healthElementDto ->
+            val (patientIdKey, _) = localCrypto.encryptAESKeyForDataOwner(
+                dataOwnerId,
+                delegateTo,
+                healthElementDto.id,
+                localCrypto.decryptEncryptionKeys(dataOwnerId, healthElementDto.cryptedForeignKeys).first()
+            )
+            val (secretForeignKey, _) = localCrypto.encryptAESKeyForDataOwner(
+                dataOwnerId,
+                delegateTo,
+                healthElementDto.id,
+                localCrypto.decryptEncryptionKeys(dataOwnerId, healthElementDto.delegations).first()
+            )
+            val (encryptionKey, _) = localCrypto.encryptAESKeyForDataOwner(
+                dataOwnerId,
+                delegateTo,
+                healthElementDto.id,
+                localCrypto.decryptEncryptionKeys(dataOwnerId, healthElementDto.encryptionKeys).first()
+            )
+
+            val delegation = DelegationDto(owner = dataOwnerId, delegatedTo = delegateTo, key = secretForeignKey)
+            val encryptionKeyDelegation =
+                DelegationDto(owner = dataOwnerId, delegatedTo = delegateTo, key = encryptionKey)
+            val cryptedForeignKeyDelegation =
+                DelegationDto(owner = dataOwnerId, delegatedTo = delegateTo, key = patientIdKey)
+
+            val delegations = healthElementDto.delegations.plus(delegateTo to setOf(delegation))
+            val encryptionKeys = healthElementDto.encryptionKeys.plus(delegateTo to setOf(encryptionKeyDelegation))
+            val cryptedForeignKeys =
+                healthElementDto.cryptedForeignKeys.plus(delegateTo to setOf(cryptedForeignKeyDelegation))
+
+            val healthElementDtoToUpdate = healthElementDto.copy(
+                delegations = delegations,
+                encryptionKeys = encryptionKeys,
+                cryptedForeignKeys = cryptedForeignKeys
+            )
+
+            try {
+                medTechApi.baseHealthElementApi.modifyHealthElement(
+                    currentUser,
+                    healthElementDtoToUpdate,
+                    ccHealthElement
+                ).toHealthcareElement()
+            } catch (e: Exception) {
+                e.printStackTrace()
+                throw IllegalStateException("Couldn't give access to dataSample")
```

Comment:
Couldn't give access to **${delegateTo} to health Element ${healthElementDtoToUpdate.id}**